sklearn/base.py

```diff
@@ -227,6 +227,12 @@ def set_params(self, **params):
                 valid_params[key] = value
 
         for key, sub_params in nested_params.items():
+            # The "base_estimator" key is special. It was deprecated and
+            # renamed to "estimator". This means we need to translate it
+            # here and set sub-parameters on "estimator", but only if the
+            # user did not explicitly set a value for "base_estimator".
+            if key == "base_estimator" and valid_params[key] == "deprecated":
+                key = "estimator"
```

Comment:
This is really tricky here:

- `base_estimator` should not be overwritten by scikit-learn
- However, we need to indicate to our users that they use a deprecated attribute that is triggered usually by modifying `base_estimator`. So the natural way would be to modify `base_estimator`

I have to think more about the right workaround.

Reply:
If we don't care about the first rule (mutate the original attribute):

```python
            if key == "base_estimator" and valid_params[key] == "deprecated":
                self.set_params(base_estimator=clone(valid_params["estimator"]))
                self.set_params(estimator=None)
                valid_params["base_estimator"] = self.base_estimator
```

This would set the estimator to raise the warning and set the parameters.

Reply:
Why wouldn't we just raise the warning in `set_params` directly instead, something like:

```python
            if key == "base_estimator" and valid_params[key] == "deprecated":
                warnings.warn(
                    f"Parameter 'base_estimator' of {self.__class__} is deprecated "
                    f"in favor of 'estimator'. See {self.__class__}'s docstring "
                    "for more details"
                )
                key = "estimator"
                valid_params[key].set_params(**sub_params)
```

Reply:
Maybe we could use a `stacklevel=2` on top of that but I am never sure about which value to use for `stacklevel` :)

Reply:
Yep, it looks simpler and it will raise fewer warnings (1 during `set_params` instead of 1 at each `fit` call). Maybe, we can restrict ourselves to scikit-learn module looking at the module of `self` to not impact 3rd party even thought the pattern `"deprecated"` indicate that 3rd party want this change.

Reply:
Implemented the warning idea from Olivier together with the restriction to estimators that are part of scikit-learn.